core/src/main/java/com/linecorp/armeria/server/ArmeriaHttpException.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server;
+
+import com.linecorp.armeria.common.HttpStatus;
+
+import io.netty.handler.codec.http2.Http2Error;
+
+/**
+ * A {@link RuntimeException} that is raised when an armeria internal http exception has occurred.
+ * This class is the general class of exceptions produced by a failed request or a reset stream.
+ */
+public abstract class ArmeriaHttpException extends RuntimeException {
+    private static final long serialVersionUID = 3487991462085151316L;
+
+    private final HttpStatus httpStatus;
+    private final Http2Error http2error;
+
+    /**
+     * Creates a new instance.
+     */
+    protected ArmeriaHttpException(HttpStatus httpStatus, Http2Error http2error) {
```

Comment:
When looking through `Http2Error`, I wasn't too sure on the semantics of `INTERNAL_ERROR` and `CANCEL` and thought they might be user-specified but if there are clear semantics for them, it makes sense to me. I've somewhat forgotten the context of filing the issue when trying to match gRPC upstream (oops!), but think it was mainly related to the HTTP/1 code, not HTTP/2